---
title: "Multi-Population Dose Response Study"
author: "Jane Park"
date: "2/6/2019"
output: 
  html_notebook: 
    fig_caption: yes
    toc: yes
  html_document: 
    keep_md: yes
    toc: yes
editor_options: 
  chunk_output_type: console
---

2019 Feb 6

Jane Park

## Libraries

```{r libraries}
library(readxl)
library(tximportData)
library(tximport)
library(lattice)
library(biomaRt)
library(readr)
library(gplots)
library(RColorBrewer)
library(edgeR)
library(limma)
library(tidyr)
library(dplyr)
library(data.table)

source("~/Documents/NIEHS_LSU_UCD/scripts/fun_MeanVar_all.R")
source("~/Documents/NIEHS_LSU_UCD/scripts/fun_TreatMeans.R")

setwd("~/Documents/NIEHS_LSU_UCD/niehs/multipop/")

```


## Import quantification tables 

Name quantification files path: 
```{r}
dir <- "~/Documents/NIEHS_LSU_UCD/niehs/multipop/salmon_quant"
files_list <- list.files(dir)
files <- file.path(dir,files_list, "quant.sf")
all(file.exists(files))
names <- gsub("_quant|_USP.*", "", files_list)
names(files) <- names
```

## Associate transcripts to gene IDs 

I'm using the gene model annotation file from supplementary materials of Noah Reid's 2016 paper, ["The genomic landscape of rapid repeated evolutionary adaptation to toxic pollution in wild fish"](http://science.sciencemag.org/content/suppl/2016/12/07/354.6317.1305.DC1)

Subset the appropriate columns from the above table. 

"We first make a data.frame called tx2gene with two columns: 1) transcript ID and 2) gene ID. The column names do not matter but this column order must be used. The transcript ID must be the same one used in the abundance files.""
```{r}
gene_names <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Fhet_gene_transcript_names.csv")
cols <- c("row", "transcript_id", "gene_id")
colnames(gene_names) <- cols
tx2gene <- gene_names[,2:3]
head(tx2gene)
```


## Convert transcript-level counts to gene-level counts, and save as a DGEList object for easy handling. Counts are scaled for library size.
txi$abundance is the TPM table for gene-level estimates. txi$counts is the counts normalized for transcript lengths. 

```{r}
txi <- tximport(files, type="salmon", tx2gene=tx2gene, countsFromAbundance = "lengthScaledTPM")
names(txi)
head(txi$counts)

y <- DGEList(txi$counts)
head(y$samples)
```

```{r Extract sample info from names}
samplenames <- rownames(y$samples)
y$samples[,"exposure"] <- NA
y$samples[,"pop"] <- NA
y$samples[,"trt"] <- NA

for (i in 1:nrow(y$samples)) {
  if (nchar(samplenames)[i] == 13) {
    y$samples$exposure[i] <- substring(samplenames[i], 9,10)
    y$samples$pop[i] <- substring(samplenames[i], 5, 7)
    
  } else if (nchar(samplenames)[i] == 12) {
    y$samples$exposure[i] <- substring(samplenames[i], 6,7)
    y$samples$pop[i] <- substring(samplenames[i], 1, 3)
    
    } else if (nchar(samplenames)[i] == 11) {
      y$samples$exposure[i] <- substring(samplenames[i],5,6)
      y$samples$pop[i] <- substring(samplenames[i], 1, 2)
      
    }
}

y$samples$trt <- with(y$samples, paste(pop, exposure, sep="."))

y$samples$exposure <- as.factor(y$samples$exposure)
y$samples$pop <- as.factor(y$samples$pop)
y$samples$trt <- as.factor(y$samples$trt)
 
```

## Filtration

```{r}
#Save unfiltered, untransformed data as another object, save as .csv file:
zz <- y	
write.table(zz$counts, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_unfiltered_counts.csv",
	col.names=TRUE, row.names=TRUE, sep="\t")
write.table(zz$samples, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_all_samples.csv",
	col.names=TRUE, row.names=TRUE, sep="\t")
```


### Filter out lowly expressed reads

Filter all samples based on counts greater than 5 in at least 4 replicate samples of 1 treatment group: 

```{r}
#Create a list of genes for which there are greater than 5 counts in at least 4 replicates for any treatment group: 
#in at least 1 treatment in at least 2 populations
d = list()
treat <- unique(y$samples$trt)
head(treat)

for (i in 1:length(treat)) {
	a <- which(y$samples$trt==treat[i])
	b <- colnames(y$counts[,a])
	kp <- rowSums(y$counts[,a] > 5) >=4 
	g <- which(kp==TRUE)
	d[[i]] <- g
}

#Subset the DGEList object y by the gene list created above:
nrow(y$counts)
filt_genes <- unique(names(unlist(d, use.names=TRUE)))
length(filt_genes)
y <- y[filt_genes,]
nrow(y$counts)

#Save filtered background gene list:
write.table(filt_genes, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_background_gene_list.csv",
	col.names=FALSE, row.names=FALSE, sep="\t")
```

## Normalize and Log Transform Read Counts

Optimal constant for rooted log transformation is around 15.

```{r}
trans <- log2(y$counts+15)
```

#### Normalize log-transformed counts for all samples: 

```{r}
mean_counts <- colMeans(trans)
norm_counts <- sweep(trans, 2, colMeans(trans), "-")
grandmean <- mean(mean_counts)
norm_counts <- norm_counts + grandmean

```

Calculate mean and standard deviation of each gene for every treatment after normalization, whole data set: 

```{r}
ttrans <- t(norm_counts)
ttrans <- as.data.frame(ttrans)
ttrans$trt <- y$samples$trt

mv <- MeanVar(ttrans)
mea <- as.vector(mv$mean)
sd <- as.vector(mv$stdev)
slope <- (coef(lm(sd~mea))[2])
print(slope)

```

Visualize mean-standard deviation relationship after normalization. 
```{r}
pdf('~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_normcountslog+15.pdf')
plot(mea,sd, pch=16, cex=0.5, main="Mean-SD relationship of Normalized Log(counts + 15), slope=-.027")
abline(coef(lm(sd~mea)),col="red",lwd=3)
lines(lowess(mea,sd,f=0.2),col="blue",lwd=3)
dev.off()

```

If everything looks ok, save normalized counts as DGEList object y. 
```{r}
y$counts <- norm_counts
```


## Unsupervised clustering of Data/ MDS plot:

```{r}
# MDS Plot
par(mfrow=c(1,1))
treats <- y$samples$trt
exp <- y$samples$exposure
pop <- y$samples$pop

col.exp <- as.factor(exp)
levels(col.exp) <- brewer.pal(nlevels(col.exp), "Set1")
col.exp <- as.character(col.exp)

col.treat <- as.factor(treats)
levels(col.treat) <- brewer.pal(nlevels(col.treat), "Set1")
col.treat <- as.character(col.treat)

col.pop <- as.factor(pop)
levels(col.pop) <- brewer.pal(nlevels(col.pop), "Set2")
col.pop <- as.character(col.pop)

plotMDS(y, labels=y$samples$trt, top=1000, col=col.pop, dim=c(1,2), gene.selection = "common", method="logFC")

w <- y[,rownames(y$samples)!="0020ARSC10351" & rownames(y$samples)!="0093ARSC00353"]
plotMDS(w, labels=w$samples$trt, top=1000, col=col.pop, dim=c(1,2), gene.selection= "common", method="logFC")
```

PCA Plot: 
```{r}
library(ggplot2)
#PCA Plot
p <- as.data.frame(t(y$counts))
p$pop <- y$samples$pop
p$exposure <- y$samples$exposure

# p <- as.data.frame(t(w$counts))
# p$pop <- w$samples$pop
# p$exposure <- w$samples$exposure

# Make PC dataframe and calculate proportion variances: 
p.pca <- prcomp(p[,1:25018])
summary(p.pca)
p.pca.proportionvariances <- ((p.pca$sdev^2) / (sum(p.pca$sdev^2)))*100
p.PCi <- data.frame(p.pca$x, pop=p$pop, exposure=p$exposure)


q <- ggplot(p.PCi, aes(x=PC1, y=PC2, color=pop, 
                              shape=exposure)) + 
  geom_point(aes(size=3)) + 
  scale_color_manual(name="Population", values=c("dodgerblue4", "deepskyblue1", "goldenrod1", "olivedrab3")) + 
  scale_shape_manual(name="Exposure", values=c(19,1,13,8)) + 
  theme_minimal() +
  xlab(paste("PC1, ", round(p.pca.proportionvariances[1], 2), "%")) + 
  ylab(paste("PC2, ", round(p.pca.proportionvariances[2], 2), "%")) +
  ggtitle("PCA of All Samples, Expression of All Background Genes ") 

# q + geom_point(aes(size=3))
#   scale_size_manual(name="Embryo Exposure", values=c(3, 3.01)) +
#   guides(fill="none", 
#          size=guide_legend(override.aes=list(shape=c(1,16))))

##Fix pca to have varying hue/saturation for exposure level. But basically, samples cluster by population, except ARS and GB, which seem to vary due to exposure

```


## Save all log-transformed & normalized counts and sample information for whole data set and subsets: 

```{r}
#Normalized counts and samples for whole data set: 
write.table(y$counts, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_allstages_all_samples_normcounts_noouts.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")
write.table(y$samples, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_allstages_all_samples_samplesinfo_noouts.csv", 
    col.names=TRUE, row.names=TRUE, sep=",")

# write.table(w$counts, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_no_outlier_normcounts_noouts.csv", 
# 		col.names=TRUE, row.names=TRUE, sep=",")
# write.table(w$samples, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_no_outlier_samplesinfo_noouts.csv", 
#     col.names=TRUE, row.names=TRUE, sep=",")

```

## DEG Analysis

```{r}
#Set up data for lm(): 
ttrans <- t(y$counts)
ttrans <- as.data.frame(ttrans)
ttrans$trt <- y$samples$trt
ttrans$exposure <- y$samples$exposure
ttrans$pop <- y$samples$pop

#Set up data frame to store p-values: 
d <- dim(ttrans)[2]-3
glist.y <- colnames(ttrans[,1:d])
cofs <- c("pop", "exposure", "pop:exposure")
pdat.y <- matrix(nrow=length(glist.y), ncol=length(cofs))
colnames(pdat.y) <- cofs
rownames(pdat.y) <- glist.y

#Run lm() on each gene: 
for (i in 1:ncol(ttrans[,1:d])) {
	g <- which(rownames(pdat.y)==colnames(ttrans)[i])
	l <- lm(formula=ttrans[,i] ~ pop*exposure, data=ttrans)
	a <- anova(l)
	p <- a$`Pr(>F)`[1:3]
	pdat.y[g,] <- p 
}

#Save p-values: 
pdat.y <- as.data.frame(pdat.y)
names(pdat.y) <- gsub(":", "_", names(pdat.y))
write.table(pdat.y, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_multipop_anova_unadj_pvalues.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")

```

#### Adjust p-values for FDR

```{r}
#Vectorize and apply p.adjust to pdat.y dataframe
pval_all <- read.csv("~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_multipop_anova_unadj_pvalues.csv", header=TRUE)

adj_pval_all <- setDT(pval_all, keep.rownames=TRUE)
colnames(adj_pval_all)[1] <- "Gene"
adj_pval_all <- adj_pval_all %>% 
	gather(`pop`, `exposure`, `pop_exposure`,
		key = "cof", value = "pvals")

adj_pval_all$adj_pvals <- p.adjust(adj_pval_all$pvals, method="fdr")

# Merge p-values with expression values for each sample: 
library(data.table)
counts <- y$counts

counts <- cbind("Gene"=rownames(counts), data.frame(counts, check.names = F, row.names=NULL))

adj_p_counts <- merge(adj_pval_all, counts, by.x="Gene")

#Write p-values and expression values for all genes: 
write.table(adj_p_counts, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_fdradj_pval_counts.csv",
		col.names=TRUE, row.names=FALSE, sep=",")

```


## Heatmaps

```{r}
# Normalize counts and average across treatments
tmv <- as.data.frame(t(y$counts))
tmv$trt <- y$samples$trt
tmv <- MeanVar(tmv)
tmv <- t(tmv$mean)
tmv <- tmv[,c("ARS.00", "ARS.10", "ARS.32", "ARS.56", 
              "GT.00", "GT.10", "GT.32", "GT.56",
              "GB.00", "GB.10", "GB.32", "GB.56",
              "VB.00", "VB.10", "VB.32", "VB.56")]
trtnorm <- as.data.frame(cbind(tmv[,1:4]-tmv[,1], tmv[,5:8]-tmv[,5], tmv[,9:12]-tmv[,9], tmv[,13:16]-tmv[,13]))
gnorm <- as.data.frame(tmv-rowMeans(tmv))

#Merge with p-values
trtnorm <- cbind("Gene"=rownames(trtnorm), data.frame(trtnorm, check.names=F, row.names=NULL))
gnorm <- cbind("Gene"=rownames(gnorm), data.frame(gnorm, check.names=F, row.names=NULL))
adj_p_trtnorm <- merge(adj_pval_all, trtnorm, by.x="Gene")
adj_p_gnorm <- merge(adj_pval_all, gnorm, by.x="Gene")

#Save averaged and normalized counts with p-values: 
write.table(adj_p_trtnorm, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_fdradj_pval_trtnorm_counts.csv",
		col.names=TRUE, row.names=FALSE, sep=",")

write.table(adj_p_gnorm, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/191101_fdradj_pval_gnorm_counts.csv",
		col.names=TRUE, row.names=FALSE, sep=",")
```


```{r}
library(pheatmap)
library(RColorBrewer)

#Main effect of exposure, excluding interaction DEGs with adj. p-value < 0.1: 
ixn <- adj_p_trtnorm[adj_p_trtnorm$cof=="pop_exposure" & 
                       adj_p_trtnorm$adj_pvals <= 0.1,]

exposure <- adj_p_trtnorm[adj_p_trtnorm$cof =="exposure" & 
                           adj_p_trtnorm$adj_pvals <= 0.05 & 
                           !(adj_p_trtnorm$Gene %in% ixn$Gene), ]

# clustering using pearson correlation
expos_main <- pheatmap(exposure[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), clustering_distance_rows="correlation", cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Embryonic Exposure", gaps_col = c(4,8,12))

exposure2 <- exposure[expos_main$tree_row$order,]

expos_main <- pheatmap(exposure2[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), clustering_distance_rows="correlation", cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Embryonic Exposure", gaps_col = c(4,8,12), border_color = NA)

#now that the hclust items all match up assign clusters to genes
exp.clust <- as.data.frame(cbind(exposure2$Gene, cluster = cutree(expos_main$tree_row, k = 2)))
exp.list <- list()
for (i in 1:length(unique(exp.clust$cluster))) {
  exp.list[[i]] <- exp.clust[exp.clust$cluster == i,]
  colnames(exp.list[[i]])[1] <- "Gene"
  rownames(exp.list[[i]]) <- c()
  names(exp.list)[i] <- paste("exp.c", i, sep="")
}

#ixn effect, clustering with pearson correlation:
ixn <- adj_p_trtnorm[adj_p_trtnorm$cof=="pop_exposure" & 
                       adj_p_trtnorm$adj_pvals <= 0.05,]

popexp <- pheatmap(ixn[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), clustering_distance_rows = "correlation", cutree_rows = 2, cluster_cols=FALSE, show_rownames = FALSE, main= "Interaction effect of Population and Embryonic Exposure", gaps_col = c(4,8,12), border_color = NA)

ixn2 <- ixn[popexp$tree_row$order,]

popexp <- pheatmap(ixn2[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), clustering_distance_rows = "correlation", cutree_rows = 2, cluster_cols=FALSE, show_rownames = FALSE, main= "Interaction effect of Population and Embryonic Exposure", gaps_col = c(4,8,12), border_color = NA)

ixn.clust <- as.data.frame(cbind(ixn2$Gene, cluster = cutree(popexp$tree_row, k = 3)))
ixn.list <- list()
for (i in 1:length(unique(ixn.clust$cluster))) {
  ixn.list[[i]] <- ixn.clust[ixn.clust$cluster == i,]
  colnames(ixn.list[[i]])[1] <- "Gene"
  rownames(ixn.list[[i]]) <- c()
  names(ixn.list)[i] <- paste("ixn.c", i, sep="")
}

#pop effect, just for fun:
ixn <- adj_p_gnorm[adj_p_gnorm$cof=="pop_exposure" & 
                       adj_p_gnorm$adj_pvals <= 0.1,]
pop <- adj_p_gnorm[adj_p_gnorm$cof =="pop" & 
                           adj_p_gnorm$adj_pvals <= 0.05 & 
                           !(adj_p_gnorm$Gene %in% ixn$Gene), ]

pop_main <- pheatmap(pop[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-1.5, to=1.5, length.out=51), cluster_cols=FALSE, clustering_distance_rows="correlation", cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Population", gaps_col = c(4,8,12))

pop2 <- pop[pop_main$tree_row$order,]

pop_main <- pheatmap(pop2[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-1, to=1.5, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Population", gaps_col = c(4,8,12))

pop.clust <- as.data.frame(cbind(pop2$Gene, cluster = cutree(pop_main$tree_row, k = 2)))
pop.list <- list()
for (i in 1:length(unique(pop.clust$cluster))) {
  pop.list[[i]] <- pop.clust[pop.clust$cluster == i,]
  colnames(pop.list[[i]])[1] <- "Gene"
  rownames(pop.list[[i]]) <- c()
  names(pop.list)[i] <- paste("pop.c", i, sep="")
}
```


## DAVID analysis set up with pheatmap clusters

```{r}
geneclusters <- list("exp"= exp.list, "ixn" = ixn.list, "pop" = pop.list)

library(readxl)
genemodels <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/Gene.Models.Conservative.and.Liberal.xlsx"))
names(genemodels) <- c("geneID", "name", "cons", "lib")
unip <- list()
length(unip) <- length(geneclusters)

for (j in 1:length(geneclusters)) {
  for (i in 1:length(geneclusters[[j]])){
  unip[[j]][[i]] <- genemodels[genemodels$geneID %in% geneclusters[[j]][[i]]$Gene,]$lib
  unip[[j]][[i]] <- unip[[j]][[i]][!is.na(unip[[j]][[i]])]
  names(unip[[j]])[i] <- names(geneclusters[[j]])[i]
  }
}
unip <- unlist(unip, recursive = FALSE, use.names = TRUE)

#Assign same length to all vectors, then run cbind: 
n <- max(lengths(unip))
for (i in 1:length(unip)) {
  length(unip[[i]]) <- n
}
david <- do.call(cbind, unip)
david[is.na(david)] <- " " 
write.table(david, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/191108_clustergenelists.txt", row.names = FALSE, col.names = TRUE, sep="\t", quote=FALSE)

#Compile background genelist for all-stages analysis by uniprot accession IDs and store one multi-list file: 

bg <- as.vector(rownames(y$counts))
bg_unip <- genemodels[genemodels$geneID %in% bg,]$lib
bg_unip <- bg_unip[!is.na(bg_unip)]
bg_unip <- as.data.frame(bg_unip)
write.table(bg_unip, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/191108_backgroundgenelists.txt", row.names=FALSE, col.names=TRUE, sep="\t", quote=FALSE)
```


# AHR pathway-related genes
```{r}
library(dplyr)
library(tidyr)
library(data.table)
library(readxl)

ahr_genes <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/ahr_target_gene_models.xlsx"))

ahr_genes <- ahr_genes[,1:4]
colnames(ahr_genes)[4] <- "Gene Annotation"

# check if they're significant for pop_exp treatment interaction
ahr_val <- adj_p_counts[adj_p_counts$Gene %in% ahr_genes$`Killifish geneID`,]
ahr_val$annotation <- NA

for (i in 1:dim(ahr_val)[1]) {
  x <- ahr_val$Gene[i]
  ahr_val$annotation[i] <- ahr_genes$`Gene Annotation`[which(ahr_genes$`Killifish geneID`==x)]
}

ahr_sig <- ahr_val[ahr_val$adj_pvals <= 0.1,]

dim(ahr_sig)

exp_ahr <- ahr_sig[which(ahr_sig$cof=="exposure"),]
pop_ahr <- ahr_sig[which(ahr_sig$cof=="pop"),]
ixn_ahr <- ahr_sig[which(ahr_sig$cof=="pop_exposure"),]
```

### Make heatmaps or expression profile figures for AHR-related genes for independent population contrast effects

# PCA of treatment groups based on normalized treatment averages 


```{r}
df <- gnorm
rownames(df) <- df$Gene
df <- df[,-1]
p <- as.data.frame(t(df))
p$pop <- as.factor(c(rep("ARS", 4), rep("GT", 4), rep("GB", 4), rep("VB", 4)))
p$pop <- factor(p$pop, levels = c("ARS", "GT", "GB", "VB"))
p$exposure <-c(rep(c("00", "10", "32", "56"), 4))
#p$trt <- y$samples$trt
p.pca <- prcomp(p[,1:25018])
summary(p.pca)
p.pca.proportionvariances <- ((p.pca$sdev^2) / (sum(p.pca$sdev^2)))*100
p.PCi <- data.frame(p.pca$x, pop=p$pop, exposure=p$exposure)

q <- ggplot(p.PCi, aes(x=PC1, y=PC2, color=pop, 
                              shape=exposure)) + 
  geom_point(aes(size=3)) + 
  scale_color_manual(name="Population", values=c("dodgerblue4", "deepskyblue1", "goldenrod1", "olivedrab3")) + 
  scale_shape_manual(name="Exposure", values=c(19,1,13,8)) + 
  theme_minimal() +
  xlab(paste("PC1, ", round(p.pca.proportionvariances[1], 2), "%")) + 
  ylab(paste("PC2, ", round(p.pca.proportionvariances[2], 2), "%")) +
  ggtitle("PCA: Treatment Averages of Expression in All Background Genes ") 

  
```

# PCA of DEG treatment averages 

### Exposure main 
```{r}
df <- gnorm
rownames(df) <- df$Gene
df <- df[,-1]
df <- df[exposure2$Gene,]
p <- as.data.frame(t(df))
p$pop <- as.factor(c(rep("ARS", 4), rep("GT", 4), rep("GB", 4), rep("VB", 4)))
p$pop <- factor(p$pop, levels = c("ARS", "GT", "GB", "VB"))
p$exposure <-as.factor(c(rep(c("00", "10", "32", "56"), 4)))
#p$trt <- y$samples$trt
p.pca <- prcomp(p[,1:4107])
summary(p.pca)
p.pca.proportionvariances <- ((p.pca$sdev^2) / (sum(p.pca$sdev^2)))*100
p.PCi <- data.frame(p.pca$x, pop=p$pop, exposure=p$exposure)

q <- ggplot(p.PCi, aes(x=PC1, y=PC2, color=pop, 
                              shape=exposure)) + 
  geom_point(aes(size=3)) + 
  scale_color_manual(name="Population", values=c("dodgerblue4", "deepskyblue1", "goldenrod1", "olivedrab3")) + 
  scale_shape_manual(name="Exposure", values=c(19,1,13,8)) + 
  theme_minimal() +
  xlab(paste("PC1, ", round(p.pca.proportionvariances[1], 2), "%")) + 
  ylab(paste("PC2, ", round(p.pca.proportionvariances[2], 2), "%")) +
  ggtitle("PCA: Treatment Averages of Expression in Exposure Main Effect Genes ") 

  
```

### Ixn effect
```{r}
df <- gnorm
rownames(df) <- df$Gene
df <- df[,-1]
df <- df[ixn2$Gene,]
p <- as.data.frame(t(df))
p$pop <- as.factor(c(rep("ARS", 4), rep("GT", 4), rep("GB", 4), rep("VB", 4)))
p$pop <- factor(p$pop, levels = c("ARS", "GT", "GB", "VB"))
p$exposure <-c(rep(c("00", "10", "32", "56"), 4))
#p$trt <- y$samples$trt
p.pca <- prcomp(p[,1:512])
summary(p.pca)
p.pca.proportionvariances <- ((p.pca$sdev^2) / (sum(p.pca$sdev^2)))*100
p.PCi <- data.frame(p.pca$x, pop=p$pop, exposure=p$exposure)

q <- ggplot(p.PCi, aes(x=PC1, y=PC2, color=pop, 
                              shape=exposure)) + 
  geom_point(aes(size=3)) + 
  scale_color_manual(name="Population", values=c("dodgerblue4", "deepskyblue1", "goldenrod1", "olivedrab3")) + 
  scale_shape_manual(name="Exposure", values=c(19,1,13,8)) + 
  theme_minimal() +
  xlab(paste("PC1, ", round(p.pca.proportionvariances[1], 2), "%")) + 
  ylab(paste("PC2, ", round(p.pca.proportionvariances[2], 2), "%")) +
  labs(title="PCA: Treatment Averages of Expression in \nPopulation x Exposure Interaction Effect Genes ") +
  theme(plot.title = element_text(hjust = 0.5))

  
```


<!-- # Heat code for MeV platform  -->
<!-- Extract read counts by DEG. Make gene lists for each treatment effect:  -->

<!-- ```{r} -->

<!-- all_cof <- list("pop"=ttrans[,as.vector(a$Gene)], "exposure"= ttrans[,as.vector(b$Gene)], "pop_exposure"= ttrans[,as.vector(c$Gene)]) -->

<!-- trtnorm <- list() -->
<!-- gnorm <- list() -->
<!-- tmv <- list() -->
<!-- for (i in 1:length(all_cof)) { -->
<!--   tmv[[i]] <- as.data.frame(all_cof[[i]]) -->
<!--   tmv[[i]]$trt <- y$samples$trt -->
<!--   tmv[[i]] <- MeanVar(tmv[[i]]) -->
<!--   tmv[[i]] <- t(tmv[[i]]$mean) -->
<!--   tmv[[i]] <- tmv[[i]][,c("ARS.00", "ARS.10", "ARS.32", "ARS.56",  -->
<!--                           "GT.00", "GT.10", "GT.32", "GT.56", -->
<!--                           "GB.00", "GB.10", "GB.32", "GB.56", -->
<!--                           "VB.00", "VB.10", "VB.32", "VB.56")] -->
<!--   trtnorm[[i]] <- as.data.frame(cbind(tmv[[i]][,1:4]-tmv[[i]][,1], tmv[[i]][,5:8]-tmv[[i]][,5], tmv[[i]][,9:12]-tmv[[i]][,9], tmv[[i]][,13:16]-tmv[[i]][,13])) -->
<!--   gnorm[[i]] <- as.data.frame(tmv[[i]]-rowMeans(tmv[[i]])) -->
<!--   write.table(trtnorm[[i]], paste("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190927_", names(all_cof[i]), "_trtnorm.txt", sep=""), col.names=NA, row.names=TRUE, sep="\t") -->
<!--   write.table(gnorm[[i]], paste("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190927_", names(all_cof[i]), "_gnorm.txt", sep=""), col.names=NA, row.names=TRUE, sep="\t") -->
<!-- } -->

<!-- names(trtnorm) <- names(all_cof) -->
<!-- names(gnorm) <- names(all_cof) -->

<!-- ``` -->



<!-- ## DAVID Analysis set up -->
<!-- After generating heatmaps and performing cluster analysis, save all clustered genes into a multi-list file for DAVID, making sure to extract only the genes with uniprot accession numbers -->

<!-- 1. List saved cluster files and import their data -->
<!-- 2. Make empty dataframe with n columns, n=number of total cluster files -->
<!-- 3. Name of cluster file - name of column, pipe funhe gene id's from each file into each column -->
<!-- 4. Make another empty dataframe with same column names and number of columns -->
<!-- 5. For every item in the column vector, if there's an entry for the uniprot accession, then write that uniprot entry into the dataframe -->
<!-- 6. Repeat above for background genelists for each stage. -->

<!-- ```{r} -->
<!-- dir <- "~/Documents/NIEHS_LSU_UCD/niehs/multipop/expression_clusters" -->
<!-- files_list <- list.files(path = dir, pattern = "190924_") -->
<!-- files <- file.path(dir, files_list) -->
<!-- filenames <- gsub("190924_", "", files_list) -->
<!-- filenames <- gsub(".txt", "", filenames) -->

<!-- #Dump cluster gene lists into a list element -->
<!-- geneclusters <- list() -->
<!-- for (i in 1:length(files)) { -->
<!--   geneclusters[[i]] <- read.csv(files[i], header=TRUE, sep = "\t") -->
<!--   colnames(geneclusters[[i]])[2] <- "Gene" -->
<!--   names(geneclusters)[i] <- filenames[i] -->
<!--   geneclusters[[i]] <- as.vector(geneclusters[[i]]$Gene) -->
<!-- } -->

<!-- #Filter above genes by those with uniprot accession number:  -->
<!-- library(readxl) -->
<!-- genemodels <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/Gene.Models.Conservative.and.Liberal.xlsx")) -->
<!-- names(genemodels) <- c("geneID", "name", "cons", "lib") -->
<!-- unip <- list() -->

<!-- for (i in 1:length(geneclusters)){ -->
<!--   unip[[i]] <- genemodels[genemodels$geneID %in% geneclusters[[i]],]$lib -->
<!--   unip[[i]] <- unip[[i]][!is.na(unip[[i]])] -->
<!--   names(unip)[i] <- names(geneclusters)[i] -->
<!-- } -->
<!-- #Assign same length to all vectors, then run cbind:  -->
<!-- n <- max(lengths(unip)) -->
<!-- for (i in 1:length(unip)) { -->
<!--   length(unip[[i]]) <- n -->
<!-- } -->
<!-- david <- do.call(cbind, unip) -->
<!-- david[is.na(david)] <- " "  -->
<!-- write.table(david, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190924_clustergenelists.txt", row.names = FALSE, col.names = TRUE, sep="\t", quote=FALSE) -->



<!-- ``` -->
