---
title: "Multi-Pop Intergenerational Study, Liver Samples"
author: "Jane Park"
date: "2020-03-09"
output: 
  html_notebook: 
    fig_caption: yes
    toc: yes
  html_document: 
    keep_md: yes
    toc: yes
editor_options: 
  chunk_output_type: console
---

```{r libraries}
library(readxl)
library(tximportData)
library(tximport)
library(lattice)
library(readr)
library(ggplot2)
library(RColorBrewer)
library(edgeR)
library(limma)
library(tidyr)
library(dplyr)

source("~/Documents/NIEHS_LSU_UCD/scripts/fun_MeanVar_all.R")
source("~/Documents/NIEHS_LSU_UCD/scripts/fun_TreatMeans.R")

setwd("~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen")

```

## Import Salmon quant data

1.Transfer quant files from farm to local machine
```{bash}

cd ~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/salmon_quant/liversamples

rsync -avzhe "ssh -p 2022" jajpark@agri.cse.ucdavis.edu:/home/jajpark/niehs/results/liver_kfish_fmd_salmcounts/* .

```

2. Import quantification files into R environment 
```{r}
library(tximportData)
library(tximport)
dir <- "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/salmon_quant/liversamples"
files_list <- list.files(dir)
files <- file.path(dir,files_list, "quant.sf")
all(file.exists(files))
names <- gsub("_quant|_USP.*", "", files_list)
names(files) <- names
```

## Associate transcripts to gene IDs 

I'm using the gene model annotation file from supplementary materials of Noah Reid's 2016 paper, ["The genomic landscape of rapid repeated evolutionary adaptation to toxic pollution in wild fish"](http://science.sciencemag.org/content/suppl/2016/12/07/354.6317.1305.DC1)

Subset the appropriate columns from the above table.

"We first make a data.frame called tx2gene with two columns: 1) transcript ID and 2) gene ID. The column names do not matter but this column order must be used. The transcript ID must be the same one used in the abundance files.""
```{r}
gene_names <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Fhet_gene_transcript_names.csv")
cols <- c("row", "transcript_id", "gene_id")
colnames(gene_names) <- cols
tx2gene <- gene_names[,2:3]
head(tx2gene)
```

## Convert transcript-level counts to gene-level counts, and save as a DGEList object for easy handling. Counts are scaled for library size.
txi$abundance is the TPM table for gene-level estimates. txi$counts is the counts normalized for transcript lengths. 

```{r}
txi <- tximport(files, type="salmon", tx2gene=tx2gene, countsFromAbundance = "lengthScaledTPM")
names(txi)
head(txi$counts)

l <- DGEList(txi$counts)
head(l$samples)
```

## Set up Sample Treatment Information

```{r sample info}
library(readxl)

sample_info <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/190903_grandis_liver_rna_JP.xlsx"))

for (i in 1:nrow(l$samples)) {
  rownames(l$samples)[i] <- substring(rownames(l$samples)[i], 1, 5)
  colnames(l$counts)[i] <- substring(colnames(l$counts)[i], 1, 5)
}

l$samples$pop <- as.factor(c(rep("VB",10), rep("GB", 10), rep("ARS", 10), rep("GT", 10)))
l$samples$exp <- as.factor(c(rep(c(rep("C",5),rep("E",5)),4)))
l$samples$trt <- as.factor(with(l$samples, paste(pop, parent, sep=".")))

```

## Filtration

Save unfiltered, untransformed data as another object, save as .csv file:
```{r save raw unfiltered counts}
l_unfil <- l 	
write.table(l_unfil$counts, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200308_unfiltered_counts.csv", col.names=TRUE, row.names=TRUE, sep="\t")

write.table(l_unfil$samples, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200308_unfiltered_samples.csv", col.names=TRUE, row.names=TRUE, sep="\t")
```


### Filter out lowly expressed reads

Filter all samples based on counts greater than 5 in at least 4 replicate samples of 1 treatment group: 

```{r filtering step}
#Create a list of genes for which there are greater than 5 counts in at least 4 replicates for each treatment group: 
d = list()
treat <- unique(l$samples$trt)
head(treat)

for (i in 1:length(treat)) {
	a <- which(l$samples$trt==treat[i])
	b <- colnames(l$counts[,a])
	kp <- rowSums(l$counts[,a] > 5) >=4 
	g <- which(kp==TRUE)
	d[[i]] <- g
}

#Subset the DGEList object y by the gene list created above:
nrow(l$counts)
filt_genes <- unique(names(unlist(d, use.names=TRUE)))
length(filt_genes)
l <- l[filt_genes,]
nrow(l$counts)

#Save filtered background gene list:
write.table(filt_genes, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200308_background_gene_list.csv", col.names=FALSE, row.names=FALSE, sep="\t")
```


## Normalize and Log Transform Read Counts

```{r}
#Transform count + constant: 
trans <- log2(l$counts+10)
```

#### Normalize log-transformed counts for all samples: 

```{r}
mean_counts <- colMeans(trans)
norm_counts <- sweep(trans, 2, colMeans(trans), "-")
grandmean <- mean(mean_counts)
norm_counts <- norm_counts + grandmean

```

Visualize mean and standard deviation relationship of each gene for every treatment after normalization, whole data set: 

```{r}
#Calculate and visualize mean and standard deviation of each gene for every treatment after normalization, whole data set: 
ttrans <- t(norm_counts)
ttrans <- as.data.frame(ttrans)
ttrans$trt <- l$samples$trt

mv <- MeanVar(ttrans)
mea <- as.vector(mv$mean)
sd <- as.vector(mv$stdev)
slope <- (coef(lm(sd~mea))[2])
print(slope)

```

Visualize mean-standard deviation relationship after normalization: 
```{r}

pdf('~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200308_normcountslog+10.pdf')
plot(mea,sd, pch=16, cex=0.5, main="Mean-SD relationship of Normalized Log(counts + 10), slope=0.005976766")
abline(coef(lm(sd~mea)),col="red",lwd=3)
lines(lowess(mea,sd,f=0.2),col="blue",lwd=3)
dev.off()
```

If everything looks ok, save normalized counts as DGEList object l. 
```{r}
l$counts <- norm_counts
```

## Unsupervised clustering of Data/ MDS plot:

```{r}
library(edgeR)
library(limma)
library(RColorBrewer)

# MDS Plot
par(mfrow=c(1,1))
exp <- l$samples$exp
pop <- l$samples$pop
treats <- l$samples$trt

col.exp <- as.factor(exp)
levels(col.exp) <-  c("red", "blue")
col.exp <- as.character(col.exp)

col.pop <- as.factor(pop)
levels(col.pop) <- brewer.pal(nlevels(col.pop), "Dark2")
col.pop <- as.character(col.pop)

col.treat <- as.factor(treats)
levels(col.treat) <- brewer.pal(nlevels(col.treat), "Set1")
col.treat <- as.character(col.treat)

plotMDS(l, labels=l$samples$trt, top=12000, col=col.pop, dim=c(1,2), gene.selection="common", method="logFC")
w <- l[,rownames(l$samples)!="LA030" & rownames(l$samples)!= "LA027"]
plotMDS(w, labels=w$samples$trt, top=500, col=col.exp, dim=c(1,2), gene.selection="common", method="logFC")

```

PCA Plot: 
```{r}
library(ggplot2)
#PCA Plot
p <- as.data.frame(t(l$counts))
p$pop <- l$samples$pop
p$exp <- l$samples$exp
p$trt <- l$samples$trt

# Make PC dataframe and calculate proportion variances: 
p.pca <- prcomp(p[,1:16781])
summary(p.pca)
p.pca.proportionvariances <- ((p.pca$sdev^2) / (sum(p.pca$sdev^2)))*100
p.PCi <- data.frame(p.pca$x, pop=p$pop, exposure=p$exp, treat=p$trt)


q <- ggplot(p.PCi, aes(x=PC1, y=PC2)) +
  geom_point(aes(colour=pop, 
                 shape=exp,
                 size=3)) + 
  scale_color_manual(name="Population", values=c("dodgerblue4", "deepskyblue1", "goldenrod1", "olivedrab3")) + 
  scale_shape_manual(name="Exposure", values=c(16,13)) + 
  theme_minimal() +
  xlab(paste("PC1, ", round(p.pca.proportionvariances[1], 2), "%")) + 
  ylab(paste("PC2, ", round(p.pca.proportionvariances[2], 2), "%")) +
  ggtitle("PCA of Liver Samples, Expression of All Background Genes ") 

```

## Save all log-transformed & normalized counts and sample information for whole data set and subsets: 

```{r}
#Normalized counts and samples for whole data set: 
write.table(l$counts, "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200310_liversamples_normcounts.csv", col.names=TRUE, row.names=TRUE, sep=",")
write.table(l$samples, "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200310_liversamples_samplesinfo.csv", col.names=TRUE, row.names=TRUE, sep=",")
```

## DEG Analysis

```{r}
#Set up data for lm(): 
ttrans <- t(l$counts)
ttrans <- as.data.frame(ttrans)
ttrans$trt <- l$samples$trt
ttrans$exp <- l$samples$exp
ttrans$pop <- l$samples$pop


#Set up data frame to store p-values: 
d <- dim(ttrans)[2]-3
glist.y <- colnames(ttrans[,1:d])
cofs <- c("pop", "exp", "pop:exp")
pdat.y <- matrix(nrow=length(glist.y), ncol=length(cofs))
colnames(pdat.y) <- cofs
rownames(pdat.y) <- glist.y

#Run lm() on each gene: 
for (i in 1:ncol(ttrans[,1:d])) {
	g <- which(rownames(pdat.y)==colnames(ttrans)[i])
	m <- lm(formula=ttrans[,i] ~ pop*exp, data=ttrans)
	a <- anova(m)
	p <- a$`Pr(>F)`[1:3]
	pdat.y[g,] <- p 
}

#Save p-values: 
pdat.y <- as.data.frame(pdat.y)
names(pdat.y) <- gsub(":", "_", names(pdat.y))
write.table(pdat.y, "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200312_anova_unadj_pvalues.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")

```


#### Adjust p-values for FDR

```{r}
library(data.table)
#Vectorize and apply p.adjust to pdat.y dataframe
#pval_all <- read.csv("~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/190925_anova_unadj_pvalues_noouts.csv", header=TRUE)
pval_all <- pdat.y

adj_pval_all <- setDT(pval_all, keep.rownames=TRUE)
colnames(adj_pval_all)[1] <- "Gene"
adj_pval_all <- adj_pval_all %>% 
	gather(`pop`, `exp`, `pop_exp`, key = "cof", value = "pvals")

adj_pval_all$adj_pvals <- p.adjust(adj_pval_all$pvals, method="fdr")

# Merge p-values with expression values for each sample: 
counts <- l$counts

counts <- cbind("Gene"=rownames(counts), data.frame(counts, check.names = F, row.names=NULL))

adj_p_counts <- merge(adj_pval_all, counts, by.x="Gene")

#Write p-values and expression values for all genes: 
write.table(adj_p_counts, "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200312_fdradj_pval_counts.csv",
		col.names=TRUE, row.names=FALSE, sep=",")
```

# Heatmaps

```{r set up datatables}
# Normalize counts and average across treatments for all genes
tmv <- as.data.frame(t(l$counts))
tmv$trt <- l$samples$trt
tmv <- MeanVar(tmv)
tmv <- t(tmv$mean)
tmv <- tmv[,c("ARS.C","ARS.E", "GT.C", "GT.E", 
              "GB.C", "GB.E", "VB.C", "VB.E")]
trtnorm <- as.data.frame(cbind(tmv[,1:2]-tmv[,1], tmv[,3:4]-tmv[,3], 
                               tmv[,5:6]-tmv[,5], tmv[,7:8]-tmv[,7]))
gnorm <- as.data.frame(tmv-rowMeans(tmv))

trtnorm <- cbind("Gene"=rownames(trtnorm), data.frame(trtnorm, check.names=F, row.names=NULL))
gnorm <- cbind("Gene"=rownames(gnorm), data.frame(gnorm, check.names=F, row.names=NULL))

#Merge with p-values
adj_p_trtnorm <- merge(adj_pval_all, trtnorm, by.x="Gene")
adj_p_gnorm <- merge(adj_pval_all, gnorm, b.x="Gene")


#Save averaged and normalized counts with p-values: 
write.table(adj_p_trtnorm, "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200312_fdradj_pval_trtnorm_counts_noarsouts.csv", col.names=TRUE, row.names=FALSE, sep=",")
write.table(adj_p_gnorm, "~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200312_fdradj_pval_gnorm_counts_noarsouts.csv", col.names=TRUE, row.names=FALSE, sep=",")

```

#### Individual population contrasts
```{r make figures: individual populations contrasts}
library(pheatmap)
library(RColorBrewer)

#Exclude interaction DEGs with adj. p-value < 0.1: 
exp_ixn <- adj_p_trtnorm[adj_p_trtnorm$cof=="pop_exp" & adj_p_trtnorm$adj_pvals <= 0.1, ]

exposure <- adj_p_trtnorm[adj_p_trtnorm$cof =="exp" & 
                           adj_p_trtnorm$adj_pvals <= 0.05 &
                            !(adj_p_trtnorm$Gene %in% exp_ixn$Gene), ]

#Make the heatmap
expos_main <- pheatmap(exposure[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Embryonic Exposure", clustering_distance_rows="correlation", gaps_col = c(2,4,6), border_color = NA)

exposure2 <- exposure[expos_main$tree_row$order,]

expos_main <- pheatmap(exposure2[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Embryonic Exposure", clustering_distance_rows="correlation", gaps_col = c(2,4,6), border_color = NA)

#Extract clusters 
exp.clust <- as.data.frame(cbind(exposure2$Gene, cluster = cutree(expos_main$tree_row, k = 2)))
exp.list <- list()
for (i in 1:length(unique(exp.clust$cluster))) {
  exp.list[[i]] <- exp.clust[exp.clust$cluster == i,]
  colnames(exp.list[[i]])[1] <- "Gene"
  rownames(exp.list[[i]]) <- c()
  names(exp.list)[i] <- paste("exp.c", i, sep="")
}


```

```{r population effect heat map}
#Exclude interaction DEGs with adj. p-value < 0.1: 
pop_ixn <- adj_p_gnorm[adj_p_gnorm$cof=="pop_exp" & adj_p_gnorm$adj_pvals <= 0.1, ]

population <- adj_p_gnorm[adj_p_gnorm$cof =="pop" & 
                           adj_p_gnorm$adj_pvals <= 0.05 &
                            !(adj_p_gnorm$Gene %in% exp_ixn$Gene), ]

#Make the heatmap
pop_main <- pheatmap(population[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Population", clustering_distance_rows="correlation", gaps_col = c(2,4,6), border_color = NA)

population2 <- population[pop_main$tree_row$order,]

pop_main <- pheatmap(population2[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Main effect of Population", clustering_distance_rows="correlation", gaps_col = c(2,4,6), border_color = NA)

#Extract clusters 
pop.clust <- as.data.frame(cbind(population2$Gene, cluster = cutree(pop_main$tree_row, k = 2)))
pop.list <- list()
for (i in 1:length(unique(pop.clust$cluster))) {
  pop.list[[i]] <- pop.clust[pop.clust$cluster == i,]
  colnames(pop.list[[i]])[1] <- "Gene"
  rownames(pop.list[[i]]) <- c()
  names(pop.list)[i] <- paste("pop.c", i, sep="")
}
```

```{r pop x exposure interaction effect heat map}
#Exclude interaction DEGs with adj. p-value < 0.1: 
popexp_ixn <- adj_p_trtnorm[adj_p_trtnorm$cof=="pop_exp" & adj_p_trtnorm$adj_pvals <= 0.1, ]

#Make the heatmap
pe_ixn <- pheatmap(popexp_ixn[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Interaction effect of Population & Exposure", clustering_distance_rows="correlation", gaps_col = c(2,4,6), border_color = NA)

popexp_ixn2 <- popexp_ixn[pe_ixn$tree_row$order,]

pe_ixn <- pheatmap(popexp_ixn2[,-(1:4)], color = colorRampPalette(c("turquoise1", "black", "yellow"))(50), breaks=seq(from=-2, to=2, length.out=51), cluster_cols=FALSE, cutree_rows = 2, show_rownames = FALSE, main= "Interaction effect of Population & Exposure", clustering_distance_rows="correlation", gaps_col = c(2,4,6), border_color = NA)

#Extract clusters 
ixn.clust <- as.data.frame(cbind(popexp_ixn2$Gene, cluster = cutree(pe_ixn$tree_row, k = 2)))
ixn.list <- list()
for (i in 1:length(unique(ixn.clust$cluster))) {
  ixn.list[[i]] <- ixn.clust[ixn.clust$cluster == i,]
  colnames(ixn.list[[i]])[1] <- "Gene"
  rownames(ixn.list[[i]]) <- c()
  names(ixn.list)[i] <- paste("ixn.c", i, sep="")
}
```

## DAVID analysis set up with pheatmap clusters

```{r}
geneclusters <- list("exp"= exp.list, "pop" = pop.list, "ixn"=ixn.list)

library(readxl)
genemodels <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/Gene.Models.Conservative.and.Liberal.xlsx"))
names(genemodels) <- c("geneID", "name", "cons", "lib")
unip <- list()
length(unip) <- length(geneclusters)

for (j in 1:length(geneclusters)) {
  for (i in 1:length(geneclusters[[j]])){
  unip[[j]][[i]] <- genemodels[genemodels$geneID %in% geneclusters[[j]][[i]]$Gene,]$lib
  unip[[j]][[i]] <- unip[[j]][[i]][!is.na(unip[[j]][[i]])]
  names(unip[[j]])[i] <- names(geneclusters[[j]])[i]
  }
}
unip <- unlist(unip, recursive = FALSE, use.names = TRUE)

#Assign same length to all vectors, then run cbind: 
n <- max(lengths(unip))
for (i in 1:length(unip)) {
  length(unip[[i]]) <- n
}
david <- do.call(cbind, unip)
david[is.na(david)] <- " " 
write.table(david, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200326_clustergenelists.txt", row.names = FALSE, col.names = TRUE, sep="\t", quote=FALSE)

#Compile background genelist for all-stages analysis by uniprot accession IDs and store one multi-list file: 

bg <- as.vector(rownames(l$counts))
bg_unip <- genemodels[genemodels$geneID %in% bg,]$lib
bg_unip <- bg_unip[!is.na(bg_unip)]
bg_unip <- as.data.frame(bg_unip)
write.table(bg_unip, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/liver_de/200326_backgroundgenelists.txt", row.names=FALSE, col.names=TRUE, sep="\t", quote=FALSE)
```


# PCA of treatment groups based on normalized treatment averages 


```{r}
df <- gnorm
rownames(df) <- df$Gene
df <- df[,-1]
p <- as.data.frame(t(df))
p$pop <- c(rep("ARS", 2), rep("GT", 2), rep("GB", 2), rep("VB", 2))
p$exposure <-c(rep(c("C","E"), 4))

dim(df)

p.pca <- prcomp(p[,1:dim(df)[1]])
summary(p.pca)
p.pca.proportionvariances <- ((p.pca$sdev^2) / (sum(p.pca$sdev^2)))*100
p.PCi <- data.frame(p.pca$x, pop=p$pop, exposure=p$exposure)

q <- ggplot(p.PCi, aes(x=PC1, y=PC2)) +
  geom_point(aes(colour=pop, 
                 shape=exposure,
                 size=3)) + 
  scale_color_manual(name="Population", values=c("dodgerblue4", "deepskyblue1", "goldenrod1", "olivedrab3")) + 
  scale_shape_manual(name="Exposure", values=c(16,13)) + 
  theme_minimal() +
  xlab(paste("PC1, ", round(p.pca.proportionvariances[1], 2), "%")) + 
  ylab(paste("PC2, ", round(p.pca.proportionvariances[2], 2), "%")) +
  ggtitle("PCA: Treatment Averages, Liver Samples ") 


```



# AHR pathway-related genes
```{r}
library(dplyr)
library(tidyr)
library(data.table)
library(readxl)

ahr_genes <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/liver_ahr_target_gene_models.xlsx"))

ahr_genes <- ahr_genes[,1:4]
colnames(ahr_genes)[4] <- "Gene Annotation"

# check if they're significant for pop_parnet treatment interaction
allcontrast_counts <- rbind(adj_p_counts, pophistory_counts, geo_counts, rand_counts)
allcontrast_counts <- geo_counts
ahr_val <- allcontrast_counts[allcontrast_counts$Gene %in% ahr_genes$`Killifish geneID`,]
ahr_val$annotation <- NA

for (i in 1:dim(ahr_val)[1]) {
  x <- ahr_val$Gene[i]
  ahr_val$annotation[i] <- ahr_genes$`Gene Annotation`[which(ahr_genes$`Killifish geneID`==x)]
}

ahr_sig <- ahr_val[ahr_val$adj_pvals <= 0.1,]

dim(ahr_sig)

ahr_sig[which(ahr_sig$cof=="geo_group" | ahr_sig$cof=="group" | ahr_sig$cof=="geo" | ahr_sig$cof=="rand_group"),]

```


